p2p/chain.py

```diff
@@ -0,0 +1,242 @@
+        while True:
+            try:
+                peer = await wait_with_token(
+                    [self._sync_requests.get()],
+                    self.cancel_token)
+            except OperationCancelled:
+                break
+
+            asyncio.ensure_future(self.sync(peer))
+
+            # TODO: If we're in light mode and we've synced up to head - 1024, trigger cancel
+            # token to stop and raise an exception to tell our caller it should perform a state
+            # sync.
+
+    async def sync(self, peer: ETHPeer) -> None:
+        if self._syncing:
+            self.logger.debug(
+                "Got a NewBlock or a new peer, but already syncing so doing nothing")
+            return
+        elif len(self._running_peers) < self.min_peers_to_sync:
+            self.logger.debug(
+                "Connected to less peers (%d) than the minimum (%d) required to sync, "
+                "doing nothing", len(self._running_peers), self.min_peers_to_sync)
+            return
+
+        self._syncing = True
+        try:
+            await self._sync(peer)
+        finally:
+            self._syncing = False
+
+    async def _sync(self, peer: ETHPeer) -> None:
+        head = await self.chaindb.coro_get_canonical_head()
+        head_td = await self.chaindb.coro_get_score(head.hash)
+        if peer.head_td <= head_td:
+            self.logger.debug(
+                "Head TD (%d) announced by %s not higher than ours (%d), not syncing",
+                peer.head_td, peer, head_td)
+            return
+
+        # FIXME: Fetch a batch of headers, in reverse order, starting from our current head, and
+        # find the common ancestor between our chain and the peer's.
+        start_at = max(0, head.block_number - eth.MAX_HEADERS_FETCH)
+        self.logger.debug("Asking %s for header batch starting at %d", peer, start_at)
+        peer.sub_proto.send_get_block_headers(start_at, eth.MAX_HEADERS_FETCH, reverse=False)
+        max_consecutive_timeouts = 3
+        consecutive_timeouts = 0
+        while True:
+            try:
+                headers = await wait_with_token(
+                    [self._new_headers.get(), peer.wait_until_finished()],
+                    self.cancel_token,
+                    timeout=3)
+            except OperationCancelled:
+                break
+            except TimeoutError:
+                self.logger.debug("Timeout waiting for header batch from %s", peer)
+                consecutive_timeouts += 1
+                if consecutive_timeouts > max_consecutive_timeouts:
+                    self.logger.debug(
+                        "Too many consecutive timeouts waiting for header batch, aborting sync "
+                        "with %s", peer)
+                    break
+                continue
+
+            if peer.is_finished():
+                self.logger.debug("%s disconnected, stopping sync", peer)
+                break
+
+            consecutive_timeouts = 0
```

Comment:
I think this timeout logic could be wrapped up in a context manager to both allow it to be tested in isolation, reused, and to reduce some of the cognitive overhead needed to parse this `while` loop.